# TP2: Machine Learning

In [1]:
import pandas as pd

# import ML packages
from sklearn.cluster import KMeans
from sklearn.datasets import make_blobs
from sklearn import linear_model
from sklearn import preprocessing

# import Plotting pckgs
from matplotlib import pyplot as plt
%matplotlib inline

In [2]:
trainingSet = pd.read_csv('../CSVs/improved_trip_train.csv')
testingSet = pd.read_csv('../CSVs/improved_trip_test.csv')
testingOriginal = pd.read_csv('../CSVs/trip_test.csv')
trainingSet.dtypes

id                     int64
duration               int64
start_date            object
start_dayOfWeek        int64
start_week             int64
start_quarter          int64
start_time            object
start_hour             int64
start_minute           int64
start_station_name    object
start_station_id       int64
end_date              object
end_dayOfWeek          int64
end_week               int64
end_quarter            int64
end_time              object
end_hour               int64
end_minute             int64
end_station_name      object
end_station_id         int64
bike_id                int64
subscription_type     object
zip_code              object
dtype: object

In [3]:
trainingSet.head()

id  duration  start_date  start_dayOfWeek  start_week  start_quarter  \
0  907649       396  2015-08-27                3          35              3   
1  384043       636  2014-07-28                0          31              3   
2  316176       334  2014-06-09                0          24              2   
3  618874       666  2015-01-26                0           5              1   
4  910977       318  2015-08-29                5          35              3   

  start_time  start_hour  start_minute  \
0   08:36:00           8            36   
1   22:06:00          22             6   
2   08:42:00           8            42   
3   16:55:00          16            55   
4   15:09:00          15             9   

                        start_station_name    ...     end_week end_quarter  \
0     Harry Bridges Plaza (Ferry Building)    ...           35           3   
1                           Market at 10th    ...           31           3   
2                        Market at Sansome    ...           24           2   
3  San Francisco Caltrain 2 (330 Townsend)    ...            5           1   
4                           Market at 10th    ...           35           3   

   end_time  end_hour  end_minute  \
0  08:43:00         8          43   
1  22:17:00        22          17   
2  08:47:00         8          47   
3  17:07:00        17           7   
4  15:14:00        15          14   

                                end_station_name  end_station_id  bike_id  \
0                         Embarcadero at Sansome              60      187   
1                           Washington at Kearny              46      417   
2                                  2nd at Folsom              62      281   
3  Temporary Transbay Terminal (Howard at Beale)              55      634   
4                             Powell Street BART              39      607   

  subscription_type  zip_code  
0        Subscriber     94602  
1        Subscriber     94133  
2        Subscriber     94107  
3        Subscriber     94602  
4        Subscriber     94709  

[5 rows x 23 columns]

In [4]:
trainingSet.shape

(549961, 23)

In [5]:
testingSet.head()

id  start_date  start_dayOfWeek  start_week  start_quarter start_time  \
0  504737  2014-10-18                5          42              4   11:25:00   
1  530846  2014-11-05                2          45              4   13:00:00   
2  813140  2015-06-18                3          25              2   17:34:00   
3  897674  2015-08-20                3          34              3   07:06:00   
4  322830  2014-06-13                4          24              2   08:46:00   

   start_hour  start_minute                        start_station_name  \
0          11            25                    Embarcadero at Sansome   
1          13             0                     Embarcadero at Folsom   
2          17            34  San Francisco Caltrain (Townsend at 4th)   
3           7             6         Civic Center BART (7th at Market)   
4           8            46   San Francisco Caltrain 2 (330 Townsend)   

   start_station_id    ...    end_week  end_quarter  end_time  end_hour  \
0                60    ...          42            4  12:12:00        12   
1                51    ...          45            4  13:09:00        13   
2                70    ...          25            2  17:37:00        17   
3                72    ...          34            3  07:15:00         7   
4                69    ...          24            2  08:57:00         8   

  end_minute               end_station_name  end_station_id bike_id  \
0         12  Powell at Post (Union Square)              71     426   
1          9      Broadway St at Battery St              82     454   
2         37                2nd at Townsend              61     370   
3         15                Townsend at 7th              65     451   
4         57          Embarcadero at Folsom              51     603   

   subscription_type  zip_code  
0           Customer     77009  
1         Subscriber     94132  
2         Subscriber     94107  
3         Subscriber     94582  
4         Subscriber     95014  

[5 rows x 22 columns]

In [6]:
trainingSet.duration.mean()

1113.4153549069842

### Let's study correlation between data

In [7]:
trainingSet.corr()['duration']

id                 -0.003133
duration            1.000000
start_dayOfWeek     0.018351
start_week          0.004002
start_quarter       0.003648
start_hour          0.002598
start_minute        0.002586
start_station_id   -0.007001
end_dayOfWeek       0.013753
end_week            0.002100
end_quarter         0.001376
end_hour            0.009626
end_minute         -0.000530
end_station_id     -0.006026
bike_id            -0.001569
Name: duration, dtype: float64

We can see that the duration is most correlated with:
    - start_dayOfWeek     0.018351
    - end_dayOfWeek       0.013753
    - end_hour            0.009626
    - start_station_id   -0.007001

In [8]:
columns = trainingSet.columns.tolist()
columns

['id',
 'duration',
 'start_date',
 'start_dayOfWeek',
 'start_week',
 'start_quarter',
 'start_time',
 'start_hour',
 'start_minute',
 'start_station_name',
 'start_station_id',
 'end_date',
 'end_dayOfWeek',
 'end_week',
 'end_quarter',
 'end_time',
 'end_hour',
 'end_minute',
 'end_station_name',
 'end_station_id',
 'bike_id',
 'subscription_type',
 'zip_code']

In [9]:
trainingSet.dtypes

id                     int64
duration               int64
start_date            object
start_dayOfWeek        int64
start_week             int64
start_quarter          int64
start_time            object
start_hour             int64
start_minute           int64
start_station_name    object
start_station_id       int64
end_date              object
end_dayOfWeek          int64
end_week               int64
end_quarter            int64
end_time              object
end_hour               int64
end_minute             int64
end_station_name      object
end_station_id         int64
bike_id                int64
subscription_type     object
zip_code              object
dtype: object

In [10]:
# Create linear regression object
regr = linear_model.LinearRegression()

In [11]:
trainingSet.shape

(549961, 23)

In [12]:
# Train the model using the training sets
regr.fit(trainingSet[['start_dayOfWeek', 'end_dayOfWeek', 'end_hour', 'start_station_id']], trainingSet.duration)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [13]:
prediction = regr.predict(testingSet[['start_dayOfWeek', 'end_dayOfWeek', 'end_hour', 'start_station_id']])

In [14]:
prediction

array([ 1603.57291184,  1069.66886328,  1298.9958606 , ...,  1784.36773396,
        1544.50655641,   596.68919234])

In [40]:
duration = { 
    'id' : testingSet.id,
    'duration' : prediction}
df_duration = pd.DataFrame(duration, columns = ['id','duration'])
df_duration = pd.merge(df_duration,testingOriginal,on = 'id',how = 'inner')
df_duration_bis = pd.DataFrame(duration, columns = ['id','duration'])
df_duration_bis = pd.merge(df_duration_bis,testingSet,on= 'id',how ='inner')
df_duration_bis = df_duration_bis.loc[:,['id','duration','start_dayOfWeek', 'end_dayOfWeek', 'end_hour', 'start_station_id']]

In [41]:
df_duration

id     duration        start_date  \
0       504737  1603.572912  10/18/2014 11:25   
1       530846  1069.668863   11/5/2014 13:00   
2       813140  1298.995861   6/18/2015 17:34   
3       897674   836.063931    8/20/2015 7:06   
4       322830  1126.487657    6/13/2014 8:46   
5       487841  1118.016993   10/7/2014 21:41   
6       677808  1150.710172   3/11/2015 18:09   
7       704449   835.598940   3/30/2015 17:29   
8       833587  1760.988055    7/5/2015 11:54   
9       420411  1534.923806   8/22/2014 13:30   
10      354777   868.292119    7/8/2014 14:05   
11       17809   883.850170   9/11/2013 13:17   
12      854667   783.069776   7/20/2015 19:45   
13      673325   311.008392     3/9/2015 8:47   
14      729900  1171.011148    4/17/2015 9:28   
15        4485  1342.114626   8/29/2013 13:16   
16      349394  1075.254624     7/3/2014 8:03   
17      784460  1342.114626   5/28/2015 13:27   
18      908431  1422.594044   8/27/2015 16:24   
19      805578  1660.769551   6/12/2015 19:57   
20      609571   356.374718   1/19/2015 12:51   
21      269385  1463.573842    5/2/2014 10:40   
22      227320   732.301734    3/26/2014 9:27   
23      658277  1353.944938   2/25/2015 17:38   
24      205479  1349.386058    3/7/2014 13:58   
25      429936  1503.354407   8/28/2014 20:09   
26      176953   839.888570    2/5/2014 14:26   
27       64226  1580.571078  10/18/2013 18:35   
28      489619  1062.225080   10/8/2014 18:36   
29      831189   986.769533     7/2/2015 8:14   
...        ...          ...               ...   
119968   60831  1106.186681  10/16/2013 17:13   
119969  875546   904.528992    8/4/2015 17:19   
119970  126578   133.757263   12/16/2013 7:48   
119971  244611  1055.699584    4/11/2014 8:22   
119972  110019  1044.247116  11/27/2013 17:14   
119973  114911  1033.712936   12/4/2013 10:54   
119974  649284  1115.597080   2/18/2015 19:45   
119975  719648   949.236528    4/10/2015 7:46   
119976  536952    27.294208   11/10/2014 6:47   
119977  579153  1105.905736  12/17/2014 16:42   
119978  619244   942.061996   1/26/2015 18:32   
119979  277802  1126.487657     5/9/2014 8:46   
119980  572524   634.406242   12/10/2014 7:06   
119981  273279   822.363902     5/6/2014 8:39   
119982  438379  1466.555645    9/4/2014 15:04   
119983  322903  1197.556676    6/13/2014 9:06   
119984  226880   869.415900   3/25/2014 17:59   
119985   34690  1087.646827   9/25/2013 14:27   
119986  581506  1412.168458  12/19/2014 17:41   
119987  365644   660.951770    7/16/2014 7:46   
119988   18373  1257.454172   9/11/2013 20:27   
119989  646921   637.949935   2/17/2015 12:59   
119990  835970   512.947026     7/7/2015 9:54   
119991  100540   862.425413  11/18/2013 18:41   
119992  503083  1126.487657   10/17/2014 8:06   
119993  274799   714.323770     5/7/2014 8:11   
119994  464786   738.827230   9/22/2014 19:38   
119995  172825  1784.367734   1/31/2014 19:03   
119996  641204  1544.506556    2/12/2015 9:25   
119997  893371   596.689192   8/17/2015 17:06   

                                   start_station_name  start_station_id  \
0                              Embarcadero at Sansome                60   
1                               Embarcadero at Folsom                51   
2            San Francisco Caltrain (Townsend at 4th)                70   
3                   Civic Center BART (7th at Market)                72   
4             San Francisco Caltrain 2 (330 Townsend)                69   
5                                     2nd at Townsend                61   
6                                      Market at 10th                67   
7                              Embarcadero at Vallejo                48   
8                                      Market at 10th                67   
9                              Embarcadero at Vallejo                48   
10                              Embarcadero at Bryant                54   
11                  Civic Center BART (7th at Market)                72  

In [42]:
df_duration_bis

id     duration  start_dayOfWeek  end_dayOfWeek  end_hour  \
0       504737  1603.572912                5              5        12   
1       530846  1069.668863                2              2        13   
2       813140  1298.995861                3              3        17   
3       897674   836.063931                3              3         7   
4       322830  1126.487657                4              4         8   
5       487841  1118.016993                1              1        21   
6       677808  1150.710172                2              2        18   
7       704449   835.598940                0              0        17   
8       833587  1760.988055                6              6        12   
9       420411  1534.923806                4              4        13   
10      354777   868.292119                1              1        14   
11       17809   883.850170                2              2        13   
12      854667   783.069776                0              0        19   
13      673325   311.008392                0              0         8   
14      729900  1171.011148                4              4         9   
15        4485  1342.114626                3              3        13   
16      349394  1075.254624                3              3         8   
17      784460  1342.114626                3              3        13   
18      908431  1422.594044                3              3        16   
19      805578  1660.769551                4              4        20   
20      609571   356.374718                0              0        12   
21      269385  1463.573842                4              4        11   
22      227320   732.301734                2              2         9   
23      658277  1353.944938                2              2        17   
24      205479  1349.386058                4              4        14   
25      429936  1503.354407                3              3        20   
26      176953   839.888570                2              2        14   
27       64226  1580.571078                4              4        18   
28      489619  1062.225080                2              2        18   
29      831189   986.769533                3              3         8   
...        ...          ...              ...            ...       ...   
119968   60831  1106.186681                2              2        17   
119969  875546   904.528992                1              1        17   
119970  126578   133.757263                0              0         7   
119971  244611  1055.699584                4              4         8   
119972  110019  1044.247116                2              2        17   
119973  114911  1033.712936                2              2        11   
119974  649284  1115.597080                2              2        19   
119975  719648   949.236528                4              4         7   
119976  536952    27.294208                0              0         6   
119977  579153  1105.905736                2              2        16   
119978  619244   942.061996                0              0        18   
119979  277802  1126.487657                4              4         8   
119980  572524   634.406242                2              2         7   
119981  273279   822.363902                1              1         8   
119982  438379  1466.555645                3              3        15   
119983  322903  1197.556676                4              4         9   
119984  226880   869.415900                1              1        18   
119985   34690  1087.646827                2              2        14   
119986  581506  1412.168458                4              4        17   
119987  365644   660.951770                2              2         7   
119988   18373  1257.454172                2              2        20   
119989  646921   637.949935                1              1        13   
119990  835970   512.947026                1              1         9 

In [ ]:
df_duration.to_csv('../CSVs/predicted_duration.csv', index=False)
df_duration_bis.to_csv('../CSVs/predicted_duration_bis.csv', index=False)